darts/models/filtering/kalman_filter.py

```diff
@@ -4,150 +4,193 @@
+            raise_if(covariates is None and self._expect_covariates,
+                    'The Kalman filter was fitted with covariates, but these were not provided.')
 
-        if not self.kf_provided:
-            kf = FpKalmanFilter(dim_x=self.dim_x, dim_z=dim_z)
-            kf.x = self.x_init
-            kf.P = self.P
-            kf.Q = self.Q
-            kf.R = self.R if self.R is not None else np.eye(dim_z)
-            kf.H = self.H if self.H is not None else np.ones((dim_z, self.dim_x))
-            kf.F = self.F
-        else:
-            raise_if_not(dim_z == self.kf.dim_z, 'The provided TimeSeries dimensionality does not match '
-                                                 'the filter observation dimensionality dim_z.')
-            kf = deepcopy(self.kf)
+            raise_if(covariates is not None and not self._expect_covariates,
+                    'Covariates were provided, but the Kalman filter was not fitted with covariates.')
 
-        super().filter(series)
-        values = series.values(copy=False)
+            raise_if_not(len(series) == len(covariates),
+                         'The length of the series must match the length of the covariates.')
+        
+            raise_if_not(covariates.is_deterministic, 'The covariates must be '
+                                                    'deterministic (observations).')
 
+        kf = deepcopy(self.kf)
+
+        y_values = series.values(copy=False)
+        if self._expect_covariates:
+            u_values = covariates.values(copy=False)
+        else:
+            u_values = np.zeros((len(y_values), 0))
+        
         # For each time step, we'll sample "n_samples" from a multivariate Gaussian
         # whose mean vector and covariance matrix come from the Kalman filter.
         if num_samples == 1:
-            sampled_states = np.zeros(((len(values)), self.dim_x, ))
+            sampled_states = np.zeros(((len(y_values)), self.dim_y, ))
         else:
-            sampled_states = np.zeros(((len(values)), self.dim_x, num_samples))
+            sampled_states = np.zeros(((len(y_values)), self.dim_y, num_samples))
 
-        # process_means = np.zeros((len(values), self.dim_x))  # mean values
-        # process_covariances = ...                            # covariance matrices; TODO
-        for i in range(len(values)):
-            obs = values[i, :]
-            kf.predict()
-            kf.update(obs)
-            mean_vec = kf.x.reshape(self.dim_x,)
+        for i in range(len(y_values)):
+            y = y_values[i, :].reshape(-1, 1)
+            u = u_values[i, :].reshape(-1, 1)
+            kf.step(y, u)
+            mean_vec = kf.y_filtereds[-1].reshape(self.dim_y,)
 
             if num_samples == 1:
-                # It's actually not sampled in this case
                 sampled_states[i, :] = mean_vec
             else:
-                cov_matrix = kf.P
+                # TODO: check formula for covariance matrix
```

Comment:
How confident are you in the formula?
It might be good to perhaps add a comment on what each of the terms are, and where the formula comes from.

Reply:
I didn't find the formula referenced at the time, but am fairly confident now that this is equivalent to (11.28) in http://web.mit.edu/kirtley/kirtley/binlustuff/literature/control/Kalman%20filter.pdf for example. I have added an explanation of the terms.